The `source-dependencies / struct` test in the `bin/run-ci-scripted-to-jar.sh` run with:

```
[error] x source-dependencies / struct
[error] sbt.internal.inc.BatchScriptRunner$PreciseScriptedError: Command failed: 'compile '
[error] Caused by: java.lang.NullPointerException
[error] 	at xsbti.api.SafeLazy$Impl.get(SafeLazy.java:58)
[error] 	at xsbti.api.Structure.declared(Structure.java:29)
[error] 	at sbt.internal.inc.binary.converters.ProtobufWriters.toStructure(ProtobufWriters.scala:283)
[error] 	at sbt.internal.inc.binary.converters.ProtobufWriters.toClassLike(ProtobufWriters.scala:535)
[error] 	at sbt.internal.inc.binary.converters.ProtobufWriters.toCompanions$1(ProtobufWriters.scala:566)
[error] 	at sbt.internal.inc.binary.converters.ProtobufWriters.toAnalyzedClass(ProtobufWriters.scala:578)
[error] 	at sbt.internal.inc.binary.converters.ProtobufWriters.$anonfun$toApis$1(ProtobufWriters.scala:669)
[error] 	at scala.collection.MapLike$MappedValues.$anonfun$foreach$3(MapLike.scala:256)
[error] 	at scala.collection.TraversableLike$WithFilter.$anonfun$foreach$1(TraversableLike.scala:792)
[error] 	at scala.collection.immutable.Map$Map3.foreach(Map.scala:195)
[error] 	at scala.collection.TraversableLike$WithFilter.foreach(TraversableLike.scala:791)
[error] 	at scala.collection.MapLike$MappedValues.foreach(MapLike.scala:256)
[error] 	at sbt.internal.inc.schema.APIs.__computeSerializedValue(APIs.scala:19)
[error] 	at sbt.internal.inc.schema.APIs.serializedSize(APIs.scala:26)
[error] 	at sbt.internal.inc.schema.APIsFile.$anonfun$writeTo$1(APIsFile.scala:40)
[error] 	at sbt.internal.inc.schema.APIsFile.$anonfun$writeTo$1$adapted(APIsFile.scala:38)
[error] 	at scala.Option.foreach(Option.scala:274)
[error] 	at sbt.internal.inc.schema.APIsFile.writeTo(APIsFile.scala:38)
[error] 	at sbt.internal.inc.binary.BinaryAnalysisFormat.writeAPIs(BinaryAnalysisFormat.scala:34)
[error] 	at sbt.internal.inc.FileAnalysisStore$BinaryFileStore.$anonfun$set$1(FileAnalysisStore.scala:91)
[error] 	at sbt.internal.inc.FileAnalysisStore$BinaryFileStore.$anonfun$set$1$adapted(FileAnalysisStore.scala:84)
[error] 	at sbt.io.Using.apply(Using.scala:22)
[error] 	at sbt.internal.inc.FileAnalysisStore$BinaryFileStore.set(FileAnalysisStore.scala:84)
[error] 	at sbt.internal.inc.AnalysisStore$CachedAnalysisStore.set(AnalysisStore.scala:27)
[error] 	at sbt.internal.inc.ProjectStructure.compile(IncHandler.scala:442)
[error] 	at sbt.internal.inc.IncHandler$$anonfun$commands$lzycompute$1.applyOrElse(IncHandler.scala:145)
```

Flaky test?  Same cause as https://github.com/sbt/zinc/issues/615?

--- a/build.sbt
+++ b/build.sbt
@@ -160,12 +160,33 @@ lazy val zinc = (project in file("zinc"))
     zincPersist,
     zincCompileCore,
     zincClassfile212,
-    zincIvyIntegration % "compile->compile;test->test",
     zincTesting % Test
   )
   .configure(addBaseSettingsAndTestDeps)
+  .enablePlugins(BuildInfoPlugin)
+  .settings(sbtbuildinfo.BuildInfoPlugin.buildInfoScopedSettings(Test))
   .settings(
     name := "zinc",
+    buildInfo in Compile := Nil, // Only generate build info for tests
+    buildInfoPackage in Test := "sbt.internal.inc",
+    buildInfoObject in Test := "ZincBuildInfo",
+    buildInfoKeys in Test := {
+      val bridgeKeys = List[BuildInfoKey](
+        BuildInfoKey.map(scalaVersion in compilerBridge210) { case (_, v) => "scalaVersion210" -> v },
+        BuildInfoKey.map(scalaInstance in compilerBridge210) { case (k, v) => "scalaJars210" -> v.allJars.toList },
+        BuildInfoKey.map(classDirectory in Compile in compilerBridge210) { case (k, v) => "classDirectory210" -> v },
+        BuildInfoKey.map(scalaVersion in compilerBridge211) { case (_, v) => "scalaVersion211" -> v },
+        BuildInfoKey.map(scalaInstance in compilerBridge211) { case (k, v) => "scalaJars211" -> v.allJars.toList },
+        BuildInfoKey.map(classDirectory in Compile in compilerBridge211) { case (k, v) => "classDirectory211" -> v },
+        BuildInfoKey.map(scalaVersion in compilerBridge212) { case (_, v) => "scalaVersion212" -> v },
+        BuildInfoKey.map(scalaInstance in compilerBridge212) { case (k, v) => "scalaJars212" -> v.allJars.toList },
+        BuildInfoKey.map(classDirectory in Compile in compilerBridge212) { case (k, v) => "classDirectory212" -> v },
+        BuildInfoKey.map(scalaVersion in compilerBridge213) { case (_, v) => "scalaVersion213" -> v },
+        BuildInfoKey.map(scalaInstance in compilerBridge213) { case (k, v) => "scalaJars213" -> v.allJars.toList },
+        BuildInfoKey.map(classDirectory in Compile in compilerBridge213) { case (k, v) => "classDirectory213" -> v },
+      )
+      bridgeKeys
+    },
     mimaSettings,
     mimaBinaryIssueFilters ++= Seq(
       exclude[DirectMissingMethodProblem]("sbt.internal.inc.IncrementalCompilerImpl.compileIncrementally"),
@@ -174,11 +195,14 @@ lazy val zinc = (project in file("zinc"))
       exclude[DirectMissingMethodProblem]("sbt.internal.inc.MixedAnalyzingCompiler.config"),
       exclude[DirectMissingMethodProblem]("sbt.internal.inc.MixedAnalyzingCompiler.makeConfig"),
       exclude[DirectMissingMethodProblem]("sbt.internal.inc.MixedAnalyzingCompiler.this"),
-      exclude[DirectMissingMethodProblem]("sbt.internal.inc.CompileConfiguration.this")
+      exclude[DirectMissingMethodProblem]("sbt.internal.inc.CompileConfiguration.this"),
+      exclude[DirectMissingMethodProblem]("sbt.internal.inc.ZincUtil.getDefaultBridgeModule"),
+      exclude[DirectMissingMethodProblem]("sbt.internal.inc.ZincUtil.scalaCompiler"),
     )
   )
 
 lazy val zincTesting = (project in internalPath / "zinc-testing")
+  .dependsOn(compilerInterface212)
   .settings(
     minimalSettings,
     noPublish,
@@ -735,7 +759,7 @@ lazy val zincClassfile212 = zincClassfileTemplate
 
 // re-implementation of scripted engine
 lazy val zincScripted = (project in internalPath / "zinc-scripted")
-  .dependsOn(zinc, zincIvyIntegration % "test->test")
+  .dependsOn(zinc % "compile;test->test")
   .dependsOn(compilerBridge210, compilerBridge211, compilerBridge212, compilerBridge213)
   .enablePlugins(BuildInfoPlugin)
   .settings(sbtbuildinfo.BuildInfoPlugin.buildInfoScopedSettings(Test))
@@ -748,26 +772,11 @@ lazy val zincScripted = (project in internalPath / "zinc-scripted")
     buildInfo in Compile := Nil,
     buildInfoPackage in Test := "sbt.internal.inc",
     buildInfoKeys in Test := {
-      val bridgeKeys = List[BuildInfoKey](
-        BuildInfoKey.map(scalaVersion in compilerBridge210) { case (_, v) => "scalaVersion210" -> v },
-        BuildInfoKey.map(scalaInstance in compilerBridge210) { case (k, v) => "scalaJars210" -> v.allJars.toList },
-        BuildInfoKey.map(classDirectory in Compile in compilerBridge210) { case (k, v) => "classDirectory210" -> v },
-        BuildInfoKey.map(scalaVersion in compilerBridge211) { case (_, v) => "scalaVersion211" -> v },
-        BuildInfoKey.map(scalaInstance in compilerBridge211) { case (k, v) => "scalaJars211" -> v.allJars.toList },
-        BuildInfoKey.map(classDirectory in Compile in compilerBridge211) { case (k, v) => "classDirectory211" -> v },
-        BuildInfoKey.map(scalaVersion in compilerBridge212) { case (_, v) => "scalaVersion212" -> v },
-        BuildInfoKey.map(scalaInstance in compilerBridge212) { case (k, v) => "scalaJars212" -> v.allJars.toList },
-        BuildInfoKey.map(classDirectory in Compile in compilerBridge212) { case (k, v) => "classDirectory212" -> v },
-        BuildInfoKey.map(scalaVersion in compilerBridge213) { case (_, v) => "scalaVersion213" -> v },
-        BuildInfoKey.map(scalaInstance in compilerBridge213) { case (k, v) => "scalaJars213" -> v.allJars.toList },
-        BuildInfoKey.map(classDirectory in Compile in compilerBridge213) { case (k, v) => "classDirectory213" -> v },
-      )
-
       Seq[BuildInfoKey](
         sourceDirectory in zinc,
         classDirectory in Test,
-        BuildInfoKey.map(dependencyClasspath in Test) { case (k, v) => "classpath" -> v.seq.map(_.data) }
-      ) ++ bridgeKeys
+        BuildInfoKey.map(dependencyClasspath in Test) { case (_, v) => "classpath" -> v.seq.map(_.data) }
+      )
     }
   )
   .configure(addSbtUtilScripted)

--- a/internal/zinc-ivy-integration/src/main/scala/sbt/internal/inc/ZincLmUtil.scala
+++ b/internal/zinc-ivy-integration/src/main/scala/sbt/internal/inc/ZincLmUtil.scala
@@ -1,0 +1,49 @@
+/*
+ * Zinc - The incremental compiler for Scala.
+ * Copyright 2011 - 2017, Lightbend, Inc.
+ * Copyright 2008 - 2010, Mark Harrah
+ * This software is released under the terms written in LICENSE.
+ */
+
+package sbt.internal.inc
+
+import java.io.File
+import java.net.URLClassLoader
+
+import sbt.librarymanagement.{ DependencyResolution, ModuleID }
+import sbt.internal.inc.classpath.ClassLoaderCache
+import xsbti._
+import xsbti.compile._
+
+object ZincLmUtil {
+  import xsbti.compile.ScalaInstance
+
+  /**
+   * Instantiate a Scala compiler that is instrumented to analyze dependencies.
+   * This Scala compiler is useful to create your own instance of incremental
+   * compilation.
+   */
+  def scalaCompiler(
+      scalaInstance: ScalaInstance,
+      classpathOptions: ClasspathOptions,
+      globalLock: GlobalLock,
+      componentProvider: ComponentProvider,
+      secondaryCacheDir: Option[File],
+      dependencyResolution: DependencyResolution,
+      compilerBridgeSource: ModuleID,
+      scalaJarsTarget: File,
+      log: Logger
+  ): AnalyzingCompiler = {
+    val compilerBridgeProvider = ZincComponentCompiler.interfaceProvider(
+      compilerBridgeSource,
+      new ZincComponentManager(globalLock, componentProvider, secondaryCacheDir, log),
+      dependencyResolution,
+      scalaJarsTarget,
+    )
+    val loader = Some(new ClassLoaderCache(new URLClassLoader(new Array(0))))
+    new AnalyzingCompiler(scalaInstance, compilerBridgeProvider, classpathOptions, _ => (), loader)
+  }
+
+  def getDefaultBridgeModule(scalaVersion: String): ModuleID =
+    ZincComponentCompiler.getDefaultBridgeModule(scalaVersion)
+}

--- a/internal/zinc-ivy-integration/src/test/scala/sbt/internal/inc/IvyBridgeProviderSpecification.scala
+++ b/internal/zinc-ivy-integration/src/test/scala/sbt/internal/inc/IvyBridgeProviderSpecification.scala
@@ -2,7 +2,6 @@ package sbt.internal.inc
 
 import java.io.File
 
-import sbt.io.IO
 import sbt.io.syntax._
 import sbt.librarymanagement._
 import sbt.librarymanagement.ivy._
@@ -14,7 +13,7 @@ import xsbti.compile.CompilerBridgeProvider
  *
  * This is a very good example on how to instantiate the compiler bridge provider.
  */
-abstract class BridgeProviderSpecification extends UnitSpec {
+abstract class IvyBridgeProviderSpecification extends UnitSpec with AbstractBridgeProviderTestkit {
   def currentBase: File = new File(".")
   def currentTarget: File = currentBase / "target" / "ivyhome"
   def currentManaged: File = currentBase / "target" / "lib_managed"
@@ -35,22 +34,6 @@ abstract class BridgeProviderSpecification extends UnitSpec {
     val manager = new ZincComponentManager(lock, componentProvider, secondaryCache, log)
     val dependencyResolution = IvyDependencyResolution(ivyConfiguration)
     ZincComponentCompiler.interfaceProvider(manager, dependencyResolution, currentManaged)
-  }
-
-  def getCompilerBridge(targetDir: File, log: Logger, scalaVersion: String): File = {
-    val provider = getZincProvider(targetDir, log)
-    val scalaInstance = provider.fetchScalaInstance(scalaVersion, log)
-    val bridge = provider.fetchCompiledBridge(scalaInstance, log)
-    val target = targetDir / s"target-bridge-$scalaVersion.jar"
-    IO.copyFile(bridge, target)
-    target
-  }
-
-  def scalaInstance(scalaVersion: String,
-                    targetDir: File,
-                    logger: Logger): xsbti.compile.ScalaInstance = {
-    val provider = getZincProvider(targetDir, logger)
-    provider.fetchScalaInstance(scalaVersion, logger)
   }
 
   private def getDefaultConfiguration(baseDirectory: File,

--- a/internal/zinc-ivy-integration/src/test/scala/sbt/internal/inc/ZincComponentCompilerSpec.scala
+++ b/internal/zinc-ivy-integration/src/test/scala/sbt/internal/inc/ZincComponentCompilerSpec.scala
@@ -3,7 +3,7 @@ package sbt.internal.inc
 import sbt.internal.util.ConsoleLogger
 import sbt.io.IO
 
-class ZincComponentCompilerSpec extends BridgeProviderSpecification {
+class ZincComponentCompilerSpec extends IvyBridgeProviderSpecification {
   val scala2105 = "2.10.5"
   val scala2106 = "2.10.6"
   val scala2118 = "2.11.8"

--- a/internal/zinc-scripted/src/test/scala/sbt/inc/MainScriptedRunner.scala
+++ b/internal/zinc-scripted/src/test/scala/sbt/inc/MainScriptedRunner.scala
@@ -4,11 +4,8 @@ import java.io.File
 
 import sbt.io.IO
 import sbt.internal.inc._
-import sbt.internal.scripted.{ CommentHandler, ScriptConfig, StatementHandler }
-import sbt.internal.util.ManagedLogger
-import xsbti.compile.CompilerBridgeProvider
-import sbt.util.Logger
 
+// WARNING: called via reflection
 final class MainScriptedRunner {
   def run(
       resourceBaseDirectory: File,
@@ -18,61 +15,8 @@ final class MainScriptedRunner {
   ): Unit = {
     IO.withTemporaryDirectory { tempDir =>
       // Create a global temporary directory to store the bridge et al
-      val handlers = new MainScriptedHandlers(tempDir, compileToJar)
+      val handlers = new IncScriptedHandlers(tempDir, compileToJar)
       ScriptedRunnerImpl.run(resourceBaseDirectory, bufferLog, tests, handlers, 4)
     }
   }
-}
-
-final class MainScriptedHandlers(tempDir: File, compileToJar: Boolean)
-    extends IncScriptedHandlers(tempDir, compileToJar) {
-  // Create a provider that uses the bridges from the classes directory of the projects
-  val provider: CompilerBridgeProvider = {
-    val compilerBridge210 = ScalaBridge(
-      BuildInfo.scalaVersion210,
-      BuildInfo.scalaJars210,
-      BuildInfo.classDirectory210
-    )
-    val compilerBridge211 = ScalaBridge(
-      BuildInfo.scalaVersion211,
-      BuildInfo.scalaJars211,
-      BuildInfo.classDirectory211
-    )
-    val compilerBridge212 = ScalaBridge(
-      BuildInfo.scalaVersion212,
-      BuildInfo.scalaJars212,
-      BuildInfo.classDirectory212
-    )
-    val compilerBridge213 = ScalaBridge(
-      BuildInfo.scalaVersion213,
-      BuildInfo.scalaJars213,
-      BuildInfo.classDirectory213
-    )
-
-    val all = List(compilerBridge210, compilerBridge211, compilerBridge212, compilerBridge213)
-    new ScriptedBridgeProvider(all, tempDir)
-  }
-
-  override def getHandlers(config: ScriptConfig): Map[Char, StatementHandler] = Map(
-    '$' -> new SleepingHandler(new ZincFileCommands(config.testDirectory()), 500),
-    '#' -> CommentHandler,
-    '>' -> {
-      val logger: ManagedLogger =
-        config.logger() match {
-          case x: ManagedLogger => x
-          case _                => sys.error("Expected ManagedLogger")
-        }
-      new BloopIncHandler(config.testDirectory(), tempDir, provider, logger, compileToJar)
-    }
-  )
-}
-
-final class BloopIncHandler(
-    directory: File,
-    cacheDir: File,
-    provider: CompilerBridgeProvider,
-    logger: ManagedLogger,
-    compileToJar: Boolean
-) extends IncHandler(directory, cacheDir, logger, compileToJar) {
-  override def getZincProvider(targetDir: File, log: Logger): CompilerBridgeProvider = provider
 }

--- a/internal/zinc-scripted/src/test/scala/sbt/internal/inc/IncScriptedRunner.scala
+++ b/internal/zinc-scripted/src/test/scala/sbt/internal/inc/IncScriptedRunner.scala
@@ -8,6 +8,7 @@ import sbt.util.Logger
 
 import scala.collection.parallel.ParSeq
 
+// WARNING: called via reflection
 class IncScriptedRunner {
   def run(
       resourceBaseDirectory: File,

--- a/internal/zinc-testing/src/main/scala/sbt/internal/inc/AbstractBridgeProviderTestkit.scala
+++ b/internal/zinc-testing/src/main/scala/sbt/internal/inc/AbstractBridgeProviderTestkit.scala
@@ -1,0 +1,24 @@
+package sbt.internal.inc
+
+import java.io.File
+import sbt.io.IO
+import sbt.io.syntax._
+import sbt.util.Logger
+import xsbti.compile.CompilerBridgeProvider
+
+trait AbstractBridgeProviderTestkit {
+  def getZincProvider(targetDir: File, log: Logger): CompilerBridgeProvider
+
+  def getCompilerBridge(targetDir: File, log: Logger, scalaVersion: String): File = {
+    val provider = getZincProvider(targetDir, log)
+    val scalaInstance = provider.fetchScalaInstance(scalaVersion, log)
+    val bridge = provider.fetchCompiledBridge(scalaInstance, log)
+    val target = targetDir / s"target-bridge-$scalaVersion.jar"
+    IO.copyFile(bridge, target)
+    target
+  }
+
+  import xsbti.compile.ScalaInstance
+  def scalaInstance(scalaVersion: String, targetDir: File, logger: Logger): ScalaInstance =
+    getZincProvider(targetDir, logger).fetchScalaInstance(scalaVersion, logger)
+}

--- a/zinc/src/main/scala/sbt/internal/inc/ZincUtil.scala
+++ b/zinc/src/main/scala/sbt/internal/inc/ZincUtil.scala
@@ -10,51 +10,38 @@ package sbt.internal.inc
 import java.io.File
 import java.net.URLClassLoader
 
-import sbt.librarymanagement.{ DependencyResolution, ModuleID }
 import sbt.internal.inc.javac.JavaTools
 import sbt.internal.inc.classpath.ClassLoaderCache
 import xsbti.compile.{ JavaTools => XJavaTools, _ }
 
-/**
- * Define a private implementation of the static methods forwarded from `ZincCompilerUtil`.
- */
+/** Define a private implementation of the static methods forwarded from `ZincCompilerUtil`. */
 object ZincUtil {
+  import xsbti.compile.ScalaInstance
 
-  /**
-   * Return a fully-fledged, default incremental compiler ready to use.
-   */
-  def defaultIncrementalCompiler: IncrementalCompiler =
-    new IncrementalCompilerImpl
+  /** Return a fully-fledged, default incremental compiler ready to use. */
+  def defaultIncrementalCompiler: IncrementalCompiler = new IncrementalCompilerImpl
 
   /**
    * Instantiate a Scala compiler that is instrumented to analyze dependencies.
    * This Scala compiler is useful to create your own instance of incremental
    * compilation.
    *
-   * @see IncrementalCompiler for more details on creating your custom
-   *     incremental compiler.
+   * @see IncrementalCompiler for more details on creating your custom incremental compiler.
    *
    * @param scalaInstance The Scala instance to be used.
-   * @param compilerBridgeJar The jar file or directory of the compiler bridge compiled for the given scala instance.
-   * @param classpathOptions The options of all the classpath that the
-   *                         compiler takes in.
+   * @param compilerBridgeJar The jar file or directory of the compiler bridge compiled for the
+   *                          given scala instance.
+   * @param classpathOptions The options of all the classpath that the compiler takes in.
    * @return A Scala compiler ready to be used.
    */
   def scalaCompiler(
-      scalaInstance: xsbti.compile.ScalaInstance,
+      scalaInstance: ScalaInstance,
       compilerBridgeJar: File,
       classpathOptions: ClasspathOptions
   ): AnalyzingCompiler = {
     val bridgeProvider = constantBridgeProvider(scalaInstance, compilerBridgeJar)
-    val emptyHandler = (_: Seq[String]) => ()
     val loader = Some(new ClassLoaderCache(new URLClassLoader(Array())))
-    new AnalyzingCompiler(
-      scalaInstance,
-      bridgeProvider,
-      classpathOptions,
-      emptyHandler,
-      loader
-    )
+    new AnalyzingCompiler(scalaInstance, bridgeProvider, classpathOptions, _ => (), loader)
   }
 
   /**
@@ -62,60 +49,19 @@ object ZincUtil {
    * This Scala compiler is useful to create your own instance of incremental
    * compilation.
    *
-   * @see IncrementalCompiler for more details on creating your custom
-   *     incremental compiler.
+   * @see IncrementalCompiler for more details on creating your custom incremental compiler.
    *
    * @param scalaInstance The Scala instance to be used.
-   * @param compilerBridgeJar The jar file or directory of the compiler bridge compiled for the given scala instance.
+   * @param compilerBridgeJar The jar file or directory of the compiler bridge compiled for the
+   *                          given scala instance.
    * @return A Scala compiler ready to be used.
    */
-  def scalaCompiler(
-      scalaInstance: xsbti.compile.ScalaInstance,
-      compilerBridgeJar: File
-  ): AnalyzingCompiler = {
-    val optionsUtil = ClasspathOptionsUtil.boot
-    scalaCompiler(scalaInstance, compilerBridgeJar, optionsUtil)
+  def scalaCompiler(scalaInstance: ScalaInstance, compilerBridgeJar: File): AnalyzingCompiler = {
+    scalaCompiler(scalaInstance, compilerBridgeJar, ClasspathOptionsUtil.boot)
   }
-
-  /**
-   * Instantiate a Scala compiler that is instrumented to analyze dependencies.
-   * This Scala compiler is useful to create your own instance of incremental
-   * compilation.
-   */
-  def scalaCompiler(
-      scalaInstance: xsbti.compile.ScalaInstance,
-      classpathOptions: ClasspathOptions,
-      globalLock: xsbti.GlobalLock,
-      componentProvider: xsbti.ComponentProvider,
-      secondaryCacheDir: Option[File],
-      dependencyResolution: DependencyResolution,
-      compilerBridgeSource: ModuleID,
-      scalaJarsTarget: File,
-      log: xsbti.Logger
-  ): AnalyzingCompiler = {
-    val componentManager =
-      new ZincComponentManager(globalLock, componentProvider, secondaryCacheDir, log)
-    val bridgeProvider =
-      ZincComponentCompiler.interfaceProvider(compilerBridgeSource,
-                                              componentManager,
-                                              dependencyResolution,
-                                              scalaJarsTarget)
-    val emptyHandler = (_: Seq[String]) => ()
-    val loader = Some(new ClassLoaderCache(new URLClassLoader(Array())))
-    new AnalyzingCompiler(
-      scalaInstance,
-      bridgeProvider,
-      classpathOptions,
-      emptyHandler,
-      loader
-    )
-  }
-
-  def getDefaultBridgeModule(scalaVersion: String): ModuleID =
-    ZincComponentCompiler.getDefaultBridgeModule(scalaVersion)
 
   def compilers(
-      instance: xsbti.compile.ScalaInstance,
+      instance: ScalaInstance,
       classpathOptions: ClasspathOptions,
       javaHome: Option[File],
       scalac: ScalaCompiler
@@ -126,7 +72,9 @@ object ZincUtil {
     Compilers.of(scalac, javaTools)
   }
 
-  def constantBridgeProvider(scalaInstance: xsbti.compile.ScalaInstance,
-                             compilerBridgeJar: File): xsbti.compile.CompilerBridgeProvider =
+  def constantBridgeProvider(
+      scalaInstance: ScalaInstance,
+      compilerBridgeJar: File,
+  ): CompilerBridgeProvider =
     ZincCompilerUtil.constantBridgeProvider(scalaInstance, compilerBridgeJar)
 }

--- a/zinc/src/test/scala/sbt/inc/ConstantBridgeProvider.scala
+++ b/zinc/src/test/scala/sbt/inc/ConstantBridgeProvider.scala
@@ -8,12 +8,15 @@ import xsbti.compile.CompilerBridgeProvider
 
 case class ScalaBridge(version: String, jars: Seq[File], classesDir: File)
 
-final class ScriptedBridgeProvider(
+final class ConstantBridgeProvider(
     bridges: List[ScalaBridge],
     tempDir: File
 ) extends CompilerBridgeProvider {
 
   import xsbti.compile.ScalaInstance
+
+  final val binSeparator = "-bin_"
+  final val javaClassVersion = System.getProperty("java.class.version")
 
   /** Get the location of the compiled Scala compiler bridge for a concrete Scala version. */
   override def fetchCompiledBridge(instance: ScalaInstance, logger: Logger): File = {
@@ -34,7 +37,6 @@ final class ScriptedBridgeProvider(
         sys.error(s"Missing $scalaVersion in supported versions ${bridges.map(_.version).mkString}")
       case Some(bridge) =>
         val targetJar: File = {
-          import sbt.internal.inc.ZincComponentCompiler.{ binSeparator, javaClassVersion }
           val id = s"scriptedCompilerBridge$binSeparator${bridge.version}__$javaClassVersion.jar"
           tempDir.toPath.resolve(id).toFile
         }

--- a/zinc/src/test/scala/sbt/inc/cached/CommonCachedCompilation.scala
+++ b/zinc/src/test/scala/sbt/inc/cached/CommonCachedCompilation.scala
@@ -58,30 +58,6 @@ abstract class CommonCachedCompilation(name: String)
 
   def remoteCacheProvider(): CacheProvider
 
-  /**
-   * Redefine what `currentManaged` is based on the project base location.
-   *
-   * The `lib_managed` folder stores all the cached libraries. It's an equivalent of
-   * `.ivy2` or `.coursier`. The analysis file provides simple read and write mappers
-   * that turn all paths relative to a **concrete position**.
-   *
-   * Note that the assumption that all paths can be made relative to a concrete position
-   * is not correct. There is no guarantee that the paths of the caches, the artifacts,
-   * the classpath entries, the outputs, et cetera share the same prefix.
-   *
-   * Such assumption is broken in our test infrastructure: `lib_managed` does not share
-   * the same prefix, and without redefining it, it fails. Note that this is a conscious
-   * design decision of the relative read and write mappers. They are focused on simplicity.
-   * Build tools that need more careful handling of paths should create their own read and
-   * write mappers.
-   *
-   * @return The file where all the libraries are stored.
-   */
-  override def currentManaged: File = {
-    import sbt.io.syntax.fileToRichFile
-    remoteProject.baseLocation.toFile./("target/lib_managed")
-  }
-
   override protected def beforeAll(): Unit = {
     //don't use temp dir because it might end up on different drive (on windows) and this later fails in sbt.internal.inc.binary.converters.ProtobufWriters due to inability to relativize paths
     val basePath =

--- a/zinc/src/test/scala/sbt/internal/inc/BridgeProviderSpecification.scala
+++ b/zinc/src/test/scala/sbt/internal/inc/BridgeProviderSpecification.scala
@@ -1,0 +1,21 @@
+package sbt.internal.inc
+
+import java.io.File
+import sbt.inc.{ ScalaBridge, ConstantBridgeProvider }
+import sbt.util.Logger
+import xsbti.compile.CompilerBridgeProvider
+
+class BridgeProviderSpecification extends UnitSpec with AbstractBridgeProviderTestkit {
+  val bridges: List[ScalaBridge] = {
+    import sbt.internal.inc.ZincBuildInfo._
+    val compilerBridge210 = ScalaBridge(scalaVersion210, scalaJars210, classDirectory210)
+    val compilerBridge211 = ScalaBridge(scalaVersion211, scalaJars211, classDirectory211)
+    val compilerBridge212 = ScalaBridge(scalaVersion212, scalaJars212, classDirectory212)
+    val compilerBridge213 = ScalaBridge(scalaVersion213, scalaJars213, classDirectory213)
+    List(compilerBridge210, compilerBridge211, compilerBridge212, compilerBridge213)
+  }
+
+  // Create a provider that uses the bridges from the classes directory of the projects
+  def getZincProvider(targetDir: File, log: Logger): CompilerBridgeProvider =
+    new ConstantBridgeProvider(bridges, targetDir)
+}

